src/Halogen/Query/HalogenQ.purs

```diff
@@ -0,0 +1,22 @@
+module Halogen.Query.HalogenQ where
+
+import Prelude
+
+import Data.Bifunctor (class Bifunctor)
+
+data HalogenQ f act i a
+  = Initialize a
+  | Finalize a
+  | Receive i a
+  | Handle act a
+  | Request (f a)
```

Comment:
Are we not doing anything wrt partiality of requests?

Reply:
I wasn't intending to - I'm not sure there's much of an advantage to it other than in the specific example case you came up with. It still needs to be dealt with somewhere, so I guess we'd have to halt with a generic message instead. At least the halt can be given a relevant message when we are forced to deal with it in the component.

Maybe I'm wrong though, and there is a better way to deal with the failed queries internally?

Reply:
How does a `halt` propagate if you are doing a child query? If I'm in a parent, and I query a child which halts, will the parent query resume with `Nothing`, or will the entire logical threat blow up?

Reply:
I see what you're saying now, I was thinking of this still causing a halt, just elsewhere - but actually it could behave much like a "failed" query does right now, in that any attempted child query might return `Nothing`.

Reply:
This does sorta seem to push me into making all `eval` partial, like `eval :: forall a. HalogenQ f act i a -> HalogenM' s act ps o m (Maybe a)` which is a bit 🤔

Reply:
That's why I added `RequestBox` in https://github.com/slamdata/purescript-halogen/issues/526#issuecomment-385271087 so that only `Request`s were potentially partial. An alternative would be to use `halt`, but I think that might be more strange, as you'd potentially be swallowing exceptions just because it happened during a request.

Reply:
Maybe you could do `Request (f a) (Unit -> a)` instead which is equivalent to the Maybe construction.

Reply:
I still haven't managed to figure out how to make this work, so I'm going to merge it as it is for now, to make a pre-release.

I see how it's supposed to work, I just haven't been able to rework eval to accommodate it. The closest I got was making `QueryBox` carry a `HalogenQ` query rather than `g`, but that falls apart due to existential `act` and `i` not working out between the `DriverState` ref and the `QueryBox`.

At least using this technique inside a HOC still works, at the expense of needing a wrapper query algebra, even if it's not baked into Halogen directly:

``` purescript
data PartialF f a = PartialF (f a) (Unit -> a)

eval = case _ of
    H.Request (PartialF q k) ->
      H.query _child unit q >>= case _ of
        Nothing -> pure (k unit)
        Just qa -> pure qa
  ...


join <$> (H.query _something unit $ HOC.PartialF (SomeQuery Just) (const Nothing))
```

Reply:
I think that's probably OK, considering partial HOCs aren't likely to be the common case.